### PR TITLE
TLD TermFrequency enhancement

### PR DESCRIPTION
**Pull request by apmoriarty**
Key Points
- function intersects docs keys prior to fetching offsets (only for TLD queries)
- document key set only considers the ROW_COLFAM (ignores keys with different column visibilities)
- greater integration of TermFrequencyConfig

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    
+    // functions may take different forms..
+    // within = {field, number, termOffsetMap, terms...}
+    // within = {number, termOffsetMap, terms...}
+    // adjacent = {field, termOffsetMap, terms...}
+    // adjacent = {termOffsetMap, terms...}
+    // phrase = {field, termOffsetMap, terms...}
+    // phrase = {termOffsetMap, terms...}
+    // scoredPhrase = {-1.5, termOffsetMap,'boy','car'}
+    private int findValueStartIndex(Function function) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: This seems like something the functions descriptor should be able to tell us.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+     * @param config
+     *            a {@link TermFrequencyConfig}
+     */
+    public DocumentKeysFunction(TermFrequencyConfig config) {
+        populateContentFunctions(config.getScript());
+    }
+    
+    protected void populateContentFunctions(JexlNode node) {
+        
+        Multimap<String,Function> functions = TermOffsetPopulator.getContentFunctions(node);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java:48]: Why only content functions?  Also the Jexl utils and the function descriptors would be more suitable.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        for (String key : functions.keySet()) {
+            Collection<Function> coll = functions.get(key);
+            for (Function f : coll) {
+                ContentFunction contentFunction = transformFunction(f);
+                contentFunctions.put(contentFunction.getField(), contentFunction);
+            }
+        }
+    }
+    
+    private ContentFunction transformFunction(Function function) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: This also seems like a weird place for this functionality.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java:48]: We're dealing with reducing the search space when aggregating term frequencies -- which have always been limited to content functions. See the TermOffsetPopulator for proof of that.

The TermOffsetPopulator call to `getContentFunctions` does make use of the JexlAstHelper which I think is the referenced jexl utility.

I'll take a look at using the content functions descriptor to extract fields and values.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: I'll take a cut at using the descriptors

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: Fair

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java (hunk 2 above)
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java:48]: Will we ever use TF offsets for anything besides content functions?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java:48]: @FineAndDandy I don't believe so

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        }
+    }
+    
+    private boolean isFunctionNegated(Function f) {
+        JexlNode node = f.args().get(0);
+        if (node != null) {
+            while (node.jjtGetParent() != null) {
+                node = node.jjtGetParent();
+                if (node instanceof ASTNotNode) {
+                    return true;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: Might we handle a not within a not?  This should travel all the way to the top of the tree.

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: JexlASTHelper.isNegated(JexlNode node) might be in order.  I am sure there are many places we could be using that.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: Switched to using the ContentFunctionsDescriptor

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java (hunk 4 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java]: Just documenting our earlier conversation: we shouldn't need special handling of negations in this case.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                    throw new IllegalArgumentException("parent was not a function node");
+                }
+                
+                argsDescriptor = descriptor.getArgumentDescriptor((ASTFunctionNode) parent);
+                
+                // content, tf, and indexed fields are not actually needed to extract fields from the function node
+                fieldsAndTerms = argsDescriptor.fieldsAndTerms(Collections.emptySet(), Collections.emptySet(), Collections.emptySet(), null);
+                
+                if (fieldsAndTerms[0].size() != 1) {
+                    throw new IllegalStateException("content function had more than one field");
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java:64]: can an ANYFIELD expansion create this? or is it always expanded into multiple functions?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java:64]: I don't believe so. At this point all fields should have been expanded and all content functions rewritten to be one field per function (that is a multi-fielded content function becomes a union of two single fielded functions).

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TFFactory.java [hunk 6]
```diff
@@ -61,9 +61,15 @@ public static com.google.common.base.Function<Tuple2<Key,Document>,Tuple3<Key,Do
...
         } else {
-            TermOffsetPopulator offsetPopulator = new TermOffsetPopulator(termFrequencyFieldValues, config.getContentExpansionFields(),
-                            config.getEvaluationFilter(), config.getSource());
-            return new TermOffsetFunction(offsetPopulator, config.getTfFields());
+            
+            DocumentKeysFunction docKeyFunction = null;
+            
+            if (config.isTld()) {
+                docKeyFunction = new DocumentKeysFunction(config);
+            }
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TFFactory.java:69]: why tld only?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TFFactory.java:69]: An event query by definition is for a single document, so the set of document keys will only be size one. 

TLD on the other hand could have many child document keys, and intersecting those before aggregating tfs will reduce the amount of work required.

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java [hunk 7]
```diff
@@ -37,20 +31,24 @@
...
+import java.util.Set;
+import java.util.SortedSet;
 
 import static datawave.query.Constants.TERM_FREQUENCY_COLUMN_FAMILY;
 
 public class TermOffsetPopulator {
     private static final Logger log = Logger.getLogger(TermOffsetPopulator.class);
     
-    private static final Set<String> phraseFunctions;
+    static final Set<String> phraseFunctions;
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java]: this can stay private

**Review by FineAndDandy — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java (hunk 7 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermOffsetPopulator.java]: good catch. Corrected this in the next commit.

**Review by ivakegg — APPROVED** [review]

**Review by FineAndDandy — APPROVED** [review]

**Pull request merged**